src/base/bittorrent/trackerentry.h

```diff
@@ -31,31 +31,38 @@
 
 #include <libtorrent/announce_entry.hpp>
 
+#include <QString>
 #include <QtGlobal>
 
-class QString;
+class QStringList;
+template <class T> class QVector;
 
 namespace BitTorrent
 {
+    class EndpointView;
+
+    enum class TrackerStatus
+    {
+        NotContacted = 1,
+        Working = 2,
+        Updating = 3,
+        NotWorking = 4,
+        WorkingWarning = 5
+    };
+
     class TrackerEntry
     {
     public:
-        enum Status
-        {
-            NotContacted = 1,
-            Working = 2,
-            Updating = 3,
-            NotWorking = 4
-        };
-
         TrackerEntry() = default;
         TrackerEntry(const QString &url);
         TrackerEntry(const lt::announce_entry &nativeEntry);
         TrackerEntry(const TrackerEntry &other) = default;
         TrackerEntry &operator=(const TrackerEntry &other) = default;
 
         QString url() const;
-        Status status() const;
+        QStringList messages() const;
```

Comment:
Getting all messages without associating them with appropriate endpoints doesn't seem useful to me. IMO, it would be better if caller side operates with endpoints directly.

Reply:
TrackerEntry class only purpose is to reduces state of all endpoints to single state. I merely follow this pattern and it works well to show the message per tracker, not only per endpoints. Moving this responsibility to caller site should be done not only for messages if anything.

Reply:
>TrackerEntry class only purpose is to reduces state of all endpoints to single state. 

Not all endpoint properties can be merged without losing their usefulness.

Reply:
>TrackerEntry class only purpose is to reduces state of all endpoints to single state.

In fact, this is just an intermediate stage of development to facilitate the transition from the previous libtorrent version, when it did not provide any per-endpoint info.

Reply:
> Not all endpoint properties can be merged without losing their usefulness.

Well it is useful to see message without expanding list of endpoints. Tracker responses are common for endpoints and those are most important for user as far as torrent is main concern. Specific endpoint errors are less important if some of them are communication with tracker.

> In fact, this is just an intermediate stage of development to facilitate the transition from the previous libtorrent version, when it did not provide any per-endpoint info.

Sure, but I'm not concerned about this "intermediate stage" in this PR. I only want to see responses from tracker, which I can't with upstream version.

Reply:
>Well it is useful to see message without expanding list of endpoints. 

So handle it at the level of logic where it (perhaps) makes some sense. I've suggested this before:
>it would be better if caller side operates with endpoints directly.

>Sure, but I'm not concerned about this "intermediate stage" in this PR. I only want to see responses from tracker, which I can't with upstream version.

I only ask you not to justify the disadvantages of your changes by existing disadvantages in the upstream. We are trying to get rid of all the obvious flaws before it is merged for the first time.

Reply:
> I only ask you not to justify the disadvantages 

I didn't justify anything. I disagree that they are flaws, except for you personal preference bias. 

There are two users of this method, both uses the data in the same way. Aggregating those messages without knowing inner logic of lt is not obvious. And would require duplicating the code. It is not just list of messages of endpoints. And `TrackerEntry` encapsulates this logic for ease of use. This is my patch and I think your proposed change is not applicable.

Reply:
>Aggregating those messages without knowing inner logic of lt is not obvious. 

There's no point in aggregating these messages at all, that's what I want to tell you. Especially in such a confusing way as you do:
If only one endpoint out of several returns a warning, what is the point of "seeing" this warning without knowing where it comes from? 
If multiple endpoints return warnings, what's the point of "seeing" them all without linking to their endpoints? 
If only one of the endpoints was working and then broke down, then you "send" a bunch of messages at once, as if all these errors happened together now - isn't that confusing?